### PR TITLE
test: extend solana unit tests

### PR DESCRIPTION
**Pull request by skosito**
# Description

- extend inbound and outbound tests for new instructions
- fix terminology, `tokenAccount` was used instead of `mintAccount`
- improve how accounts are resolved in inbound instructions

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Release Notes

- **New Features**
	- Introduced a new struct for SPL withdrawal instructions, enhancing functionality with a decimal precision field.
	- Added tests for SPL withdrawal message hashing and instruction parsing, improving validation and error handling.

- **Improvements**
	- Streamlined variable naming from `tokenAccount` to `mintAccount` across multiple components for clarity.
	- Enhanced deposit instruction handling with improved account resolution and error messaging.

- **Documentation**
	- Added a detailed JSON file documenting an outbound transaction on the Solana blockchain.

### CODE REVIEW

File: pkg/contracts/solana/gateway_message_test.go [hunk 1]
```diff
@@ -47,3 +47,24 @@ func Test_MsgWhitelistHash(t *testing.T) {
...
+		toAta, _, err := solana.FindAssociatedTokenAddress(to, mintAccount)
+		require.NoError(t, err)
+
+		wantHash := "87fa5c0ed757c6e1ea9d8976537eaf7868bc1f1bbf55ab198a01645d664fe0ae"
+		wantHashBytes, err := hex.DecodeString(wantHash)
+		require.NoError(t, err)
+
+		// create new withdraw message
+		hash := contracts.NewMsgWithdrawSPL(chainID, nonce, amount, 8, mintAccount, to, toAta).Hash()
+		require.True(t, bytes.Equal(hash[:], wantHashBytes))
```
Comment (lumtis) [inline pkg/contracts/solana/gateway_message_test.go:68]: doesn't `require.EqualValues` work for bytes? This allows to see the diff on failure

File: pkg/contracts/solana/gateway_message_test.go [hunk 2]
```diff
@@ -47,3 +47,24 @@ func Test_MsgWhitelistHash(t *testing.T) {
 		require.True(t, bytes.Equal(hash[:], wantHashBytes))
 	})
 }
+
+func Test_MsgWithdrawSPLHash(t *testing.T) {
+	t.Run("should pass for archived inbound, receipt and cctx", func(t *testing.T) {
```
Comment (lumtis) [inline pkg/contracts/solana/gateway_message_test.go:52]: Can we make the test desc clearer?

File: pkg/contracts/solana/gateway_message_test.go [hunk 3]
```diff
@@ -47,3 +47,24 @@ func Test_MsgWhitelistHash(t *testing.T) {
...
+		// #nosec G115 always positive
```
Comment (lumtis) [inline pkg/contracts/solana/gateway_message_test.go:53]: Let's use the ARRANGE-ACT-ASSERT pattern we have in comments in other tests. Same for other tests in the PR

File: pkg/contracts/solana/inbound.go [hunk 4]
```diff
@@ -94,34 +94,53 @@ func ParseInboundAsDepositSPL(
...
+	if len(instructionAccounts) != accountsNumDeposit {
+		return "", fmt.Errorf("want %d accounts, got %d", accountsNumDeposit, len(instructionAccounts))
 	}
 
-	// sender is the signer account
-	return tx.Message.AccountKeys[0].String(), nil
+	// the accounts are [signer, pda, system_program]
+	// check if first account is signer
+	if !instructionAccounts[0].IsSigner {
+		return "", fmt.Errorf("unexpected signer %s", instructionAccounts[0].PublicKey.String())
```
Comment (lumtis) [inline pkg/contracts/solana/inbound.go:112]: Desc should rather say "not signer" than "unexpected signer"?

File: pkg/contracts/solana/inbound_test.go [hunk 5]
```diff
@@ -67,6 +67,61 @@ func Test_ParseInboundAsDeposit(t *testing.T) {
...
+			Discriminator: DiscriminatorDepositSPL,
+			Memo:          inst.Memo,
+		})
+		require.NoError(t, err)
+
+		tx.Message.Instructions[0].Data = data
+
+		deposit, err := ParseInboundAsDeposit(tx, 0, txResult.Slot)
+		require.NoError(t, err)
+		require.Nil(t, deposit)
```
Comment (lumtis) [inline pkg/contracts/solana/inbound_test.go:95]: I don't this assertion is necessary. deposit should always be ignored if err is returned in any case

**Review by lumtis — APPROVED** [review]

File: pkg/contracts/solana/gateway_message_test.go [hunk 6]
```diff
@@ -42,8 +46,36 @@ func Test_MsgWhitelistHash(t *testing.T) {
...
+		chainID := uint64(chains.SolanaLocalnet.ChainId)
+		nonce := uint64(0)
+		amount := uint64(1336000)
+		mintAccount := solana.MustPublicKeyFromBase58("AS48jKNQsDGkEdDvfwu1QpqjtqbCadrAq9nGXjFmdX3Z")
+		to := solana.MustPublicKeyFromBase58("37yGiHAnLvWZUNVwu9esp74YQFqxU1qHCbABkDvRddUQ")
+		toAta, _, err := solana.FindAssociatedTokenAddress(to, mintAccount)
+		require.NoError(t, err)
+
+		wantHash := "87fa5c0ed757c6e1ea9d8976537eaf7868bc1f1bbf55ab198a01645d664fe0ae"
+		wantHashBytes, err := hex.DecodeString(wantHash)
```
Comment (ws4charlie) [inline pkg/contracts/solana/gateway_message_test.go:71]: You could use `wantHashBytes := testutil.HexToBytes(t, wantHash)` for simplification if you want. Same for other occurrences.

File: pkg/contracts/solana/inbound.go [hunk 7]
```diff
@@ -94,34 +94,53 @@ func ParseInboundAsDepositSPL(
...
-		return "", fmt.Errorf("want %d accounts, got %d", accountsNumDeposit, len(inst.Accounts))
+	if len(instructionAccounts) != accountsNumDeposit {
+		return "", fmt.Errorf("want %d accounts, got %d", accountsNumDeposit, len(instructionAccounts))
 	}
 
-	// sender is the signer account
-	return tx.Message.AccountKeys[0].String(), nil
+	// the accounts are [signer, pda, system_program]
+	// check if first account is signer
+	if !instructionAccounts[0].IsSigner {
```
Comment (ws4charlie) [inline pkg/contracts/solana/inbound.go:111]: looks nice!

File: pkg/contracts/solana/instruction.go [hunk 8]
```diff
@@ -128,6 +128,7 @@ type WithdrawSPLInstructionParams struct {
 	// Discriminator is the unique identifier for the withdraw instruction
 	Discriminator [8]byte
 
+	// Decimals is decimals for spl token
```
Comment (ws4charlie) [inline pkg/contracts/solana/instruction.go:131]: Should we remove this field? I saw it's being read nowhere.

**Review by ws4charlie — APPROVED** [review]
looks good. nit comment

**Pull request merged**